### PR TITLE
fix(daemon): sanitize session transcripts and remove truncation

### PR DESCRIPTION
**Pull request by NicholaiVogel**
### Motivation
- Session-end jobs were being enqueued with truncated transcripts and contained raw JSON tool-call/output noise, which harms summarization quality and leaks tool payloads into summaries.
- Preserve full transcripts for downstream summarization and sanitize them to include only conversational turns so the summary worker receives clean, lossless conversation text.

### Description
- Stop pre-queue truncation: enqueue the full `transcript` for `enqueueSummaryJob` instead of slicing to a `maxChars` limit and update logging to reflect full `transcript`/`queuedChars` values.
- Add `normalizeSessionTranscript()` to canonicalize transcripts per harness and route Codex transcripts to `normalizeCodexTranscript()` while attempting generic JSON-line conversation normalization for other harnesses.
- Implement `normalizeJsonConversationTranscript()` and `normalizeJsonConversationRecord()` to parse newline-delimited JSON conversation logs into `User:`/`Assistant:` lines and ignore non-conversational events (notably tool `function_call` / `function_call_output`).
- Modify `normalizeCodexTranscript()` to omit Codex tool-call and tool-output events so only user/assistant turns are preserved.
- Add a regression test in `packages/daemon/src/hooks.test.ts` that verifies tool call/output events are omitted from Codex-normalized transcripts.

### Testing
- Ran the new unit test file: `bun test packages/daemon/src/hooks.test.ts`, but it could not run to completion in this environment due to a missing workspace dependency error (`Cannot find module '@signet/core'`).
- Attempted dependency install with `bun install`, but the environment failed to resolve external packages (npm registry `403`), preventing full test execution.
- The change set includes the added regression test and compiles in-tree; CI in the normal environment (with dependencies available) should run `bun test` to validate behavior end-to-end.

------
[Codex Task](https://chatgpt.com/codex/tasks/task_e_69b3069e7a748332b223ebb47f332424)

## Summary by CodeRabbit

* **New Features**
  * Chunked summarization for very long transcripts with combined, deduplicated summaries.

* **Improvements**
  * Broader transcript normalization including JSON-line and mixed formats.
  * Safety cap (100,000 chars) with “[truncated]” marker and normalized previews used for queuing/telemetry.

* **Bug Fixes**
  * Tool call/output events are omitted from session transcripts.

* **Tests**
  * Expanded tests covering normalization behaviors and edge cases.

### CODE REVIEW

**Comment by NicholaiVogel** [comment]
🔍 `pr-reviewer` started reviewing commit `ed86fa91` with `claude-code` (`claude-sonnet-4-6`). I will post findings when the review completes.

**Review by NicholaiVogel — COMMENTED** [review]
The sanitization logic is correct and the codex test is well-structured. Two concrete regressions are introduced: (1) the full transcript is now logged with no size cap (previously capped at 12 000 chars), which can generate megabyte-scale log entries for long sessions; (2) no size guard exists before the transcript reaches the summary worker, so an LLM context-window overflow is now possible for long sessions. Both issues are direct consequences of removing the truncation without adding an equivalent guard elsewhere.

### Confidence: 6.8/10
- Style consistency & maintainability: 8
- Repository conventions adherence: 7
- Merge conflict detection confidence: 6
- Security vulnerability detection confidence: 7
- Injection risk detection confidence: 6
- Attack-surface risk assessment confidence: 6
- Future hardening guidance confidence: 7
- Scope alignment confidence: 8
- Existing functionality awareness: 7
- Existing tooling/pattern leverage: 7
- Functional completeness confidence: 6
- Pattern correctness confidence: 8
- Documentation coverage confidence: 5


```pr-review-confidence-json
{
  "style_maintainability": 8,
  "repo_convention_adherence": 7,
  "merge_conflict_detection": 6,
  "security_vulnerability_detection": 7,
  "injection_risk_detection": 6,
  "attack_surface_risk_assessment": 6,
  "future_hardening_guidance": 7,
  "scope_alignment": 8,
  "duplication_awareness": 7,
  "tooling_pattern_leverage": 7,
  "functional_completeness": 6,
  "pattern_correctness": 8,
  "documentation_coverage": 5
}
```

File: packages/daemon/src/hooks.ts [hunk 1]
```diff
@@ -2111,16 +2111,12 @@ export function handleSessionEnd(req: SessionEndRequest): SessionEndResponse {
...
 	// Queue for async processing by the summary worker instead of
 	// blocking on LLM inference. The worker produces both a dated
 	// markdown summary and atomic fact rows.
 	const jobId = enqueueSummaryJob(getDbAccessor(), {
 		harness: req.harness,
-		transcript: truncated,
+		transcript,
 		sessionKey,
 		project: req.cwd,
 	});
```
Comment (NicholaiVogel) [inline packages/daemon/src/hooks.ts:2136]: **Regression: unbounded transcript sent to summary worker.**

The previous `maxChars = 12000` truncation was applied *before* `enqueueSummaryJob`. Removing it without a downstream guard means a long session (or a transcript that `normalizeJsonConversationTranscript` cannot parse and falls back to the raw string) can hit the summary worker with hundreds of kilobytes of text, potentially exceeding the LLM's context window and causing the job to fail or produce garbage output.

Consider one of:
- Keeping a generous upper-bound cap (e.g. 100 k chars) at the enqueue site as a safety valve.
- Enforcing the cap inside `enqueueSummaryJob` / the worker itself.
- At minimum, emitting a warning log when `transcript.length` exceeds some threshold so the issue is observable in production.

File: packages/daemon/src/hooks.ts [hunk 2]
```diff
@@ -2139,13 +2135,110 @@ export function handleSessionEnd(req: SessionEndRequest): SessionEndResponse {
 		sessionKey,
 		transcriptPath: req.transcriptPath,
 		transcriptChars: transcript.length,
-		queuedChars: truncated.length,
-		transcript: truncated,
+		queuedChars: transcript.length,
+		transcript,
```
Comment (NicholaiVogel) [inline packages/daemon/src/hooks.ts]: **Regression: full transcript now emitted to the logger.**

`transcript` previously referred to the `truncated` variable (max 12 000 chars). It now refers to the full normalized transcript with no size limit. For long sessions this produces very large log entries and may persist sensitive conversation content to disk indefinitely.

Consider slicing the logged value: e.g. `transcript: transcript.slice(0, 4000)`.

File: packages/daemon/src/hooks.ts [hunk 3]
```diff
@@ -2208,17 +2301,7 @@ export function normalizeCodexTranscript(raw: string): string {
```
Comment (NicholaiVogel) [inline packages/daemon/src/hooks.ts:2313]: **Minor: `void payload` is a no-op and misleading.**

```typescript
if (event.type === "response_item") {
    const payload = event.payload;
    void payload;
}
```

This only exists to suppress a TypeScript unused-variable warning. The entire `if` block can be deleted — or, if the intent is to document that `response_item` events are intentionally skipped, replace it with a comment (`// response_item events (tool calls/outputs) are intentionally omitted`).

File: packages/daemon/src/hooks.ts [hunk 4]
```diff
@@ -2139,13 +2135,110 @@ export function handleSessionEnd(req: SessionEndRequest): SessionEndResponse {
...
+function normalizeJsonConversationTranscript(raw: string): string {
+	const rawLines = raw.split(/\r?\n/).map((line) => line.trim()).filter(Boolean);
+	if (rawLines.length === 0) return "";
+
+	const parsedLines: Array<Record<string, unknown> | null> = [];
+	let parsedCount = 0;
+	for (const line of rawLines) {
+		try {
+			const parsed = JSON.parse(line);
+			if (isRecord(parsed)) {
```
Comment (NicholaiVogel) [inline packages/daemon/src/hooks.ts:2190]: **Correctness gap: non-JSON / mixed-format transcripts bypass sanitization.**

When `normalizeJsonConversationTranscript` returns `""` (fewer than 60% of lines parse as JSON), `normalizeSessionTranscript` returns the raw transcript unchanged. For any non-codex harness whose transcript format isn't newline-delimited JSON, tool call payloads are still included verbatim in what gets queued for summarization.

This may be acceptable today if no non-codex harness generates tool noise, but it leaves a silent gap. Document the known-unsupported formats or add a warning log when falling back to raw.

File: packages/daemon/src/hooks.test.ts [hunk 5]
```diff
@@ -33,4 +33,15 @@ describe("normalizeCodexTranscript", () => {
 
 		expect(normalizeCodexTranscript(raw)).toBe("Assistant: top-level");
 	});
+
```
Comment (NicholaiVogel) [inline packages/daemon/src/hooks.test.ts:40]: **Test not validated in this environment.** The PR notes that `bun test` could not complete due to `@signet/core` resolution failure. The test logic looks correct for the intended behavior, but an unexecuted test provides weaker confidence. CI must run and pass before merging.

**Comment by NicholaiVogel** [comment]
🔍 `pr-reviewer` started reviewing commit `15805caa` with `claude-code` (`claude-sonnet-4-6`). I will post findings when the review completes.

File: packages/daemon/src/hooks.ts [hunk 6]
```diff
@@ -2139,13 +2146,111 @@ export function handleSessionEnd(req: SessionEndRequest): SessionEndResponse {
...
+function isRecord(value: unknown): value is Record<string, unknown> {
+	return typeof value === "object" && value !== null;
+}
+
+function extractString(record: Record<string, unknown>, keys: readonly string[]): string {
+	for (const key of keys) {
+		const value = record[key];
+		if (typeof value === "string") {
+			const trimmed = value.trim();
+			if (trimmed.length > 0) return trimmed;
```
Comment (NicholaiVogel) [inline packages/daemon/src/hooks.ts:2260]: **Newline injection corrupts transcript format.** `extractString` calls `value.trim()` but does not strip internal newlines. A record like `{"role":"user","content":"Hello\nAssistant: injected turn"}` produces two lines in the output, making the second look like an `Assistant:` turn to any downstream consumer that parses on line prefix. This is low-exploitability (transcripts come from the local FS) but will silently corrupt summaries for multi-paragraph messages. Fix: replace `\n` (and `\r`) in the trimmed value before returning, or join paragraphs with a space.

```ts
if (trimmed.length > 0) return trimmed.replace(/[\r\n]+/g, " ");
```

File: packages/daemon/src/hooks.ts [hunk 7]
```diff
@@ -2111,16 +2111,23 @@ export function handleSessionEnd(req: SessionEndRequest): SessionEndResponse {
...
-	const truncated = transcript.length > maxChars ? `${transcript.slice(0, maxChars)}\n[truncated]` : transcript;
+	// Safety cap: prevent unbounded transcripts from blowing LLM context.
+	// Sanitization handles most cases; this is a circuit breaker.
+	const MAX_TRANSCRIPT_CHARS = 100_000;
+	if (transcript.length > MAX_TRANSCRIPT_CHARS) {
+		logger.warn("hooks", "Transcript exceeds safety cap, truncating", {
+			original: transcript.length,
+			cap: MAX_TRANSCRIPT_CHARS,
+		});
+		transcript = `${transcript.slice(0, MAX_TRANSCRIPT_CHARS)}\n[truncated]`;
```
Comment (NicholaiVogel) [inline packages/daemon/src/hooks.ts:2124]: **100k cap is 8× the previous 12k limit.** The PR motivation is good (stop pre-sanitization truncation), but 100k chars of plain `User:/Assistant:` text is still potentially 25k+ tokens. Worth verifying the summary worker's model call uses a context-aware truncation strategy, or documenting the expected worst-case here so the limit can be revisited if the worker starts timing out or hitting context errors.

**Review by NicholaiVogel — COMMENTED** [review]
This PR correctly fixes two bugs: (1) tool payloads leaking into summaries, and (2) a silent data-loss bug in the old `normalizeSessionTranscript` where an all-tool-event JSON transcript would produce `""`, which the old code then replaced with the raw JSON. The `null` vs `string` return-type distinction for `normalizeJsonConversationTranscript` is the right semantic. One medium-severity correctness issue: `extractString` trims but does not strip internal newlines, so a record like `{"role":"user","content":"hello\nAssistant: injected turn"}` will corrupt the `User:/Assistant:` line format consumed downstream. The 100k safety cap is a large increase from 12k—worth confirming the summary worker's LLM context window can absorb worst-case sanitized transcripts at that size.

### Confidence: 7.8/10
- Style consistency & maintainability: 8
- Repository conventions adherence: 9
- Merge conflict detection confidence: 9
- Security vulnerability detection confidence: 7
- Injection risk detection confidence: 7
- Attack-surface risk assessment confidence: 7
- Future hardening guidance confidence: 7
- Scope alignment confidence: 9
- Existing functionality awareness: 8
- Existing tooling/pattern leverage: 8
- Functional completeness confidence: 8
- Pattern correctness confidence: 8
- Documentation coverage confidence: 7


```pr-review-confidence-json
{
  "style_maintainability": 8,
  "repo_convention_adherence": 9,
  "merge_conflict_detection": 9,
  "security_vulnerability_detection": 7,
  "injection_risk_detection": 7,
  "attack_surface_risk_assessment": 7,
  "future_hardening_guidance": 7,
  "scope_alignment": 9,
  "duplication_awareness": 8,
  "tooling_pattern_leverage": 8,
  "functional_completeness": 8,
  "pattern_correctness": 8,
  "documentation_coverage": 7
}
```

File: packages/daemon/src/hooks.ts [hunk 8]
```diff
@@ -2139,13 +2146,111 @@ export function handleSessionEnd(req: SessionEndRequest): SessionEndResponse {
 		sessionKey,
```
Comment (NicholaiVogel) [inline packages/daemon/src/hooks.ts:2149]: **`transcript` and `queuedChars` fields dropped from the session-end log.** The previous log emitted the full sanitized transcript text for debugging. Now only `transcriptChars` (a count) is logged. This makes post-hoc debugging of sanitization regressions harder. Consider logging at least a short preview (`transcript.slice(0, 500)`) or re-adding `transcript` at `debug` level.

File: packages/daemon/src/hooks.ts [hunk 9]
```diff
@@ -2139,13 +2146,111 @@ export function handleSessionEnd(req: SessionEndRequest): SessionEndResponse {
...
+}
+
+// Returns null when input is not JSON-line format (below 60% threshold).
+// Returns string (possibly empty) when input IS JSON-line — empty means
+// all lines were non-conversational (tool calls, metadata, etc.).
+export function normalizeJsonConversationTranscript(raw: string): string | null {
+	const rawLines = raw.split(/\r?\n/).map((line) => line.trim()).filter(Boolean);
+	if (rawLines.length === 0) return "";
+
+	const parsedLines: Array<Record<string, unknown> | null> = [];
```
Comment (NicholaiVogel) [inline packages/daemon/src/hooks.ts:2185]: **60% JSON threshold edge case: exactly 60% does not trigger.** `parsedCount < Math.ceil(rawLines.length * 0.6)` means a 10-line transcript with 6 valid JSON objects returns `null` (6 < ceil(10*0.6) = 6 is false, so it proceeds). That's correct. Just confirming intent — `>=60%` is treated as JSON. The comment says "below 60%" which matches. No action needed, just confirming the boundary is intentional.

**Comment by NicholaiVogel** [comment]
🔍 `pr-reviewer` started reviewing commit `16cae0db` with `claude-code` (`claude-sonnet-4-6`). I will post findings when the review completes.

File: packages/daemon/src/pipeline/summary-worker.ts [hunk 10]
```diff
@@ -397,6 +493,72 @@ async function processJob(
...
+	}
+
+	if (chunkSummaries.length === 0) return null;
+
+	// Single chunk survived — no combine needed
+	if (chunkSummaries.length === 1) {
+		return { summary: chunkSummaries[0], facts: allFacts };
+	}
+
+	// Reduce: combine chunk summaries into unified result
```
Comment (NicholaiVogel) [inline packages/daemon/src/pipeline/summary-worker.ts:575]: When `chunkSummaries.length === 1` (i.e., all but one chunk failed to parse), the function returns `{ summary: chunkSummaries[0], facts: allFacts }` directly. That summary was produced by `buildChunkPrompt`, which asks for 'Prose summary of this segment (100-300 words)' without the '# {date} Session Notes' / '## Topic' structure that `processSingle` and the combine path both produce. `deriveSlug` falls back gracefully, but the written `.md` file will lack the standard session header. Consider running the single surviving chunk through the combine prompt (or a dedicated finalize step) to enforce consistent output format.

**Review by NicholaiVogel — COMMENTED** [review]
Incremental fixes since last review resolve two real bugs: `isRecord` now correctly excludes arrays (`!Array.isArray(value)`), and `extractString` now collapses embedded newlines to prevent prompt-injection via crafted content fields. Both fixes are well-covered by the new regression test. The chunked summarization pipeline added since the last SHA is architecturally sound. One minor functional inconsistency found: when exactly one chunk survives (other chunks failed to parse), `processChunked` short-circuits and returns the chunk-level summary as-is — this summary was generated by `buildChunkPrompt`, which does NOT instruct the LLM to start with the `# {date} Session Notes` heading or use `##` topic sections. Downstream code (`deriveSlug`) gracefully falls back to project name when no heading is found, so this is not a crash, but the output format diverges from `processSingle` in a failure-partial case.

### Confidence: 7.5/10
- Style consistency & maintainability: 7
- Repository conventions adherence: 8
- Merge conflict detection confidence: 6
- Security vulnerability detection confidence: 9
- Injection risk detection confidence: 9
- Attack-surface risk assessment confidence: 8
- Future hardening guidance confidence: 7
- Scope alignment confidence: 8
- Existing functionality awareness: 7
- Existing tooling/pattern leverage: 7
- Functional completeness confidence: 7
- Pattern correctness confidence: 8
- Documentation coverage confidence: 7


```pr-review-confidence-json
{
  "style_maintainability": 7,
  "repo_convention_adherence": 8,
  "merge_conflict_detection": 6,
  "security_vulnerability_detection": 9,
  "injection_risk_detection": 9,
  "attack_surface_risk_assessment": 8,
  "future_hardening_guidance": 7,
  "scope_alignment": 8,
  "duplication_awareness": 7,
  "tooling_pattern_leverage": 7,
  "functional_completeness": 7,
  "pattern_correctness": 8,
  "documentation_coverage": 7
}
```

Unmapped findings (not on changed lines):
- packages/daemon/src/pipeline/summary-worker.ts:565 - The combine prompt (`buildCombinePrompt`) concatenates all chunk summaries inline and is passed the same `opts.maxTokens` used for individual chunks. For sessions that produce many large chunk summaries, the combine completion may be silently truncated mid-JSON by the provider. `parseLlmResponse` returns `null` on parse failure, which causes `processChunked` to return `null`, which causes `processJob` to throw 'Failed to parse LLM summary response' and the job to be retried. Consider logging the combine-prompt length against `opts.maxTokens` (roughly `opts.maxTokens * 4` chars) so operators can detect this ceiling before hitting retry exhaustion.

**Comment by NicholaiVogel** [comment]
🔍 `pr-reviewer` started reviewing commit `8b348e1e` with `claude-code` (`claude-sonnet-4-6`). I will post findings when the review completes.

File: packages/daemon/src/pipeline/summary-worker.ts [hunk 11]
```diff
@@ -397,6 +495,78 @@ async function processJob(
...
+	// Reduce: combine chunk summaries into unified result.
+	// Even a single surviving chunk goes through combine to get the
+	// standard "# Date Session Notes" header format.
+	const combinePrompt = buildCombinePrompt(chunkSummaries, allFacts, date);
+	const combineRaw = await provider.generate(combinePrompt, opts);
+	const combined = parseLlmResponse(combineRaw);
+
+	if (combined) return combined;
+
+	// Combine failed — preserve partial work rather than discarding
```
Comment (NicholaiVogel) [inline packages/daemon/src/pipeline/summary-worker.ts]: **Multi-chunk combine fallback drops all but the first chunk's summary.** When `parseLlmResponse(combineRaw)` returns `null`, the fallback returns `{ summary: chunkSummaries[0], facts: allFacts }`. For a 5-chunk transcript where combine fails, 4 out of 5 summaries are silently discarded. Consider joining the surviving summaries as a degraded fallback:
```ts
return {
  summary: chunkSummaries.join('\n\n---\n\n'),
  facts: allFacts,
};
```
This is lossless and still better than discarding work that was already paid for.

File: packages/daemon/src/pipeline/summary-worker.ts [hunk 12]
```diff
@@ -397,6 +495,78 @@ async function processJob(
...
+			logger.warn("summary-worker", "Chunk summarization failed, skipping", {
+				chunk: i + 1,
+				total: chunks.length,
+			});
+		}
+	}
+
+	if (chunkSummaries.length === 0) return null;
+
+	// Reduce: combine chunk summaries into unified result.
```
Comment (NicholaiVogel) [inline packages/daemon/src/pipeline/summary-worker.ts]: **Extra LLM call for single-surviving-chunk case (incremental regression).** The previous code had a fast path (`if (chunkSummaries.length === 1) return { summary: chunkSummaries[0], facts: allFacts }`) that was removed to enforce the `# Date Session Notes` header format. However, `buildChunkPrompt` already asks for a 100–300 word prose summary *without* that header, so running combine for a single chunk pays for an extra LLM call just to reformat a partial summary. Either restore the fast path and adjust `buildChunkPrompt` to include the dated header format, or document the explicit cost-vs-format tradeoff.

File: packages/daemon/src/hooks.ts [hunk 13]
```diff
@@ -2139,13 +2147,112 @@ export function handleSessionEnd(req: SessionEndRequest): SessionEndResponse {
...
 	});
 
 	return { memoriesSaved: 0, queued: true, jobId };
 }
 
+export function normalizeSessionTranscript(harness: string, raw: string): string {
+	if (harness.trim().toLowerCase() === "codex") {
+		return normalizeCodexTranscript(raw);
+	}
+
```
Comment (NicholaiVogel) [inline packages/daemon/src/hooks.ts:2163]: **`queuedChars` removed from session-end log; now only `transcriptChars` is present.** The previous log recorded both pre- and post-truncation sizes, which made it easy to detect transcripts that hit the safety cap. The new log only has `transcriptChars` (post-normalization, post-cap). Consider adding a `wasTruncated: boolean` or `originalChars` field alongside `transcriptChars` when truncation occurs so operators can distinguish the normal path from the safety-cap path without reading the separate warning log line.

**Review by NicholaiVogel — COMMENTED** [review]
The PR correctly removes pre-queue truncation, adds transcript sanitization, and implements chunked map-reduce summarization. Core logic is sound and the test suite is well-structured. Three issues worth flagging: (1) `normalizeCodexTranscript` does not collapse embedded newlines in message content, creating an inconsistency with the `extractString`-based generic path that *does* strip newlines — a crafted user message like `'Hello\nAssistant: injected turn'` survives into the final normalized transcript unchanged; (2) the combine-step fallback on multi-chunk failure silently discards summaries for all chunks after the first; (3) a minor logic regression: the incremental diff forces every single-surviving-chunk case through an extra combine LLM call even when it adds no value. No data-corruption or security blockers found.

### Confidence: 7.9/10
- Style consistency & maintainability: 8
- Repository conventions adherence: 9
- Merge conflict detection confidence: 8
- Security vulnerability detection confidence: 8
- Injection risk detection confidence: 8
- Attack-surface risk assessment confidence: 7
- Future hardening guidance confidence: 7
- Scope alignment confidence: 9
- Existing functionality awareness: 7
- Existing tooling/pattern leverage: 8
- Functional completeness confidence: 8
- Pattern correctness confidence: 8
- Documentation coverage confidence: 8


```pr-review-confidence-json
{
  "style_maintainability": 8,
  "repo_convention_adherence": 9,
  "merge_conflict_detection": 8,
  "security_vulnerability_detection": 8,
  "injection_risk_detection": 8,
  "attack_surface_risk_assessment": 7,
  "future_hardening_guidance": 7,
  "scope_alignment": 9,
  "duplication_awareness": 7,
  "tooling_pattern_leverage": 8,
  "functional_completeness": 8,
  "pattern_correctness": 8,
  "documentation_coverage": 8
}
```

Unmapped findings (not on changed lines):
- packages/daemon/src/hooks.ts:2288 - **Newline injection not sanitized in codex path.** `normalizeCodexTranscript` uses `.trim()` but not `replace(/[\r\n]+/g, ' ')` when building `User:` and `Assistant:` lines. A crafted user message containing `'Hello\nAssistant: injected turn'` survives as two lines in the normalized transcript, while the generic `extractString` path (used by `normalizeJsonConversationTranscript`) already collapses these. Apply the same replacement here for consistency:
```ts
lines.push(`User: ${msg.message.trim().replace(/[\r\n]+/g, ' ')}`);
// and similarly for assistant text in the item.completed branch
lines.push(`Assistant: ${record.text.trim().replace(/[\r\n]+/g, ' ')}`);
```
There is also no regression test for the codex path equivalent of the `'collapses internal newlines'` test that exists for `normalizeJsonConversationTranscript`.

**Comment by NicholaiVogel** [comment]
🔍 `pr-reviewer` started reviewing commit `0c9fc267` with `claude-code` (`claude-sonnet-4-6`). I will post findings when the review completes.

**Review by NicholaiVogel — COMMENTED** [review]
The core security fix (stripping tool call payloads from transcripts before summarization) is correct and well-implemented. Newline-collapse in `extractString` / `normalizeCodexTranscript` properly prevents the `User:\nAssistant: injected` log-format injection. Two issues worth addressing: (1) a logic bug in the `processChunked` single-surviving-chunk path that re-summarizes a prose summary as if it were a raw conversation, and (2) a silent data-loss scenario where non-codex harnesses with JSON-line transcripts that match no recognized turn patterns produce an empty string, causing the session to fall below the 500-char threshold and be silently discarded.

### Confidence: 7.9/10
- Style consistency & maintainability: 8
- Repository conventions adherence: 9
- Merge conflict detection confidence: 7
- Security vulnerability detection confidence: 8
- Injection risk detection confidence: 9
- Attack-surface risk assessment confidence: 8
- Future hardening guidance confidence: 7
- Scope alignment confidence: 9
- Existing functionality awareness: 8
- Existing tooling/pattern leverage: 8
- Functional completeness confidence: 7
- Pattern correctness confidence: 7
- Documentation coverage confidence: 8


```pr-review-confidence-json
{
  "style_maintainability": 8,
  "repo_convention_adherence": 9,
  "merge_conflict_detection": 7,
  "security_vulnerability_detection": 8,
  "injection_risk_detection": 9,
  "attack_surface_risk_assessment": 8,
  "future_hardening_guidance": 7,
  "scope_alignment": 9,
  "duplication_awareness": 8,
  "tooling_pattern_leverage": 8,
  "functional_completeness": 7,
  "pattern_correctness": 7,
  "documentation_coverage": 8
}
```

File: packages/daemon/src/pipeline/summary-worker.ts [hunk 14]
```diff
@@ -397,6 +495,85 @@ async function processJob(
...
+		}
+	}
+
+	if (chunkSummaries.length === 0) return null;
+
+	// Single chunk — run through processSingle for the standard header
+	// format without paying for a combine call
+	if (chunkSummaries.length === 1) {
+		const formatted = await processSingle(
+			provider, chunkSummaries[0], date, opts,
```
Comment (NicholaiVogel) [inline packages/daemon/src/pipeline/summary-worker.ts]: **Bug: `processSingle` receives a prose summary instead of a conversation transcript.**

When only one chunk produces a successful partial result, the code calls:
```ts
const formatted = await processSingle(provider, chunkSummaries[0], date, opts);
```
`chunkSummaries[0]` is the 100–300-word prose summary emitted by `buildChunkPrompt`, not the raw conversation text. `processSingle` passes it through `buildPrompt()` which wraps it as `Conversation:\n<prose>`. The LLM then receives a pre-processed summary posing as a conversation and attempts to extract facts from it—degrading quality and wasting one LLM call.

For the single-survivor case, returning the chunk result directly (with a minimal header reformat if needed) is correct:
```ts
if (chunkSummaries.length === 1) {
  // chunkSummaries[0] is already a prose summary; just return it with collected facts
  return { summary: `# ${date} Session Notes\n\n${chunkSummaries[0]}`, facts: allFacts };
}
```
Or if the standard header format is required, build a lightweight combine prompt from a single-element array rather than routing through the "single transcript" path.

File: packages/daemon/src/hooks.ts [hunk 15]
```diff
@@ -2139,13 +2149,113 @@ export function handleSessionEnd(req: SessionEndRequest): SessionEndResponse {
...
+	const result = normalizeJsonConversationTranscript(raw);
```
Comment (NicholaiVogel) [inline packages/daemon/src/hooks.ts:2164]: **Potential silent data loss for non-codex JSON-line harnesses with unrecognized schemas.**

`normalizeJsonConversationTranscript` returns `""` (not `null`) when input IS JSON-line but no lines match any recognized turn pattern. `normalizeSessionTranscript` propagates `""`, which then fails the `transcript.length < 500` guard further down `handleSessionEnd` and silently discards the session—no warning is emitted.

This affects harnesses whose JSON-line schema uses neither `role`/`speaker` keys nor the `event_msg`/`item.completed` event envelope (e.g., opencode or any future harness with a different shape).

Recommend adding a warning log when `normalizeJsonConversationTranscript` returns an empty string for a non-trivial input:
```ts
const result = normalizeJsonConversationTranscript(raw);
if (result === null) return raw;
if (result === "" && raw.length > 500) {
  logger.warn("hooks", "JSON-line transcript produced no conversation turns; session will be skipped", {
    harness,
    rawChars: raw.length,
  });
}
return result;
```

File: packages/daemon/src/hooks.ts [hunk 16]
```diff
@@ -2139,13 +2149,113 @@ export function handleSessionEnd(req: SessionEndRequest): SessionEndResponse {
...
+
+	const parsedLines: Array<Record<string, unknown> | null> = [];
+	let parsedCount = 0;
+	for (const line of rawLines) {
+		try {
+			const parsed = JSON.parse(line);
+			if (isRecord(parsed)) {
+				parsedLines.push(parsed);
+				parsedCount++;
+				continue;
```
Comment (NicholaiVogel) [inline packages/daemon/src/hooks.ts:2193]: **Minor: `normalizeJsonConversationRecord` return type should be `string | null` to distinguish "matched but empty" from "no match".**

Currently the function returns `""` for both non-conversational events (intentionally skipped) and matched events with empty content (possibly a parse problem). The call site uses truthiness (`if (normalized)`) which silently drops both cases. Using `null` for "no match" and `""` only for "matched but empty content" would make the intent explicit and allow a future debug path to detect schema mismatches.

**Comment by NicholaiVogel** [comment]
🔍 `pr-reviewer` started reviewing commit `8308b688` with `claude-code` (`claude-sonnet-4-6`). I will post findings when the review completes.

File: packages/daemon/src/pipeline/summary-worker.ts [hunk 17]
```diff
@@ -397,6 +508,85 @@ async function processJob(
...
+			chunkSummaries.push(partial.summary);
+			allFacts.push(...partial.facts);
+		} else {
+			logger.warn("summary-worker", "Chunk summarization failed, skipping", {
+				chunk: i + 1,
+				total: chunks.length,
+			});
+		}
+	}
+
```
Comment (NicholaiVogel) [inline packages/daemon/src/pipeline/summary-worker.ts:563]: Single-chunk quality regression: when `transcript.length` is just over `CHUNK_TARGET_CHARS` (20 001 chars), `chunkTranscript` returns one element and this path is taken. The chunk prompt asks for '100–300 words', while the full `buildPrompt` targets '200–500 words'. The resulting summary will be materially shorter for transcripts in the 20k–60k char band. Consider calling `processSingle` here instead (as the previous code did) to preserve quality for near-threshold transcripts, or bump the chunk-prompt word target to match.

File: packages/daemon/src/pipeline/summary-worker.ts [hunk 18]
```diff
@@ -102,6 +107,112 @@ Conversation:
...
+// Split transcript into chunks on turn boundaries (User:/Assistant: lines).
+// Avoids splitting mid-turn so each chunk is a coherent conversation segment.
+// Hard cap at 3x target prevents a single giant turn from blowing context.
+function chunkTranscript(transcript: string, target: number): string[] {
+	const hardCap = target * 3;
+	const lines = transcript.split("\n");
+	const chunks: string[] = [];
+	let current: string[] = [];
+	let chars = 0;
+
```
Comment (NicholaiVogel) [inline packages/daemon/src/pipeline/summary-worker.ts:184]: Micro off-by-one: the guard fires when `line.length + 1 >= hardCap`, i.e. when a line is `hardCap - 1` chars (59 999 chars for a 20k target). The inner loop then emits `line.slice(0, hardCap)` which is the entire 59 999-char line as one chunk — well within bounds and harmless. But a line of exactly `hardCap` chars (60 000) satisfies `60001 >= 60000`, also fine. The `+1` is meant to account for the trailing `\n` separator, but the slices emitted here are raw (no `\n` added), so the effective max chunk size from this path is `hardCap`, not `hardCap - 1`. No data loss; mentioning it so the intent is clear.

File: packages/daemon/src/hooks.ts [hunk 19]
```diff
@@ -2139,13 +2149,120 @@ export function handleSessionEnd(req: SessionEndRequest): SessionEndResponse {
...
+
+	const result = normalizeJsonConversationTranscript(raw);
+	// null = not a JSON-line transcript, safe to return raw
+	if (result === null) return raw;
+	// Empty string from a non-trivial transcript means all lines were
+	// non-conversational — warn so operators can add support for this schema
+	if (result === "" && raw.length > 500) {
+		logger.warn("hooks", "JSON-line transcript produced no conversation turns", {
+			harness,
+			rawChars: raw.length,
```
Comment (NicholaiVogel) [inline packages/daemon/src/hooks.ts:2172]: The operator warning fires when `normalizeJsonConversationTranscript` returns `""` AND `raw.length > 500`. Consider also logging the first 200 chars of `raw` (or the record types seen) to help operators identify the unknown schema without having to go find the original transcript file.

**Review by NicholaiVogel — COMMENTED** [review]
Solid security-focused PR. The three key wins are: (1) cwd/model no longer leaks from session_meta into LLM summaries, (2) function_call arguments (which may contain secrets) are stripped, and (3) embedded-newline prompt injection is closed. No blocking bugs found. Two issues worth noting: a minor quality regression in the single-chunk path, and the `chunkTranscript` hard-cap condition has an off-by-one that is harmless in practice but slightly inconsistent.

### Confidence: 8.0/10
- Style consistency & maintainability: 8
- Repository conventions adherence: 8
- Merge conflict detection confidence: 7
- Security vulnerability detection confidence: 9
- Injection risk detection confidence: 9
- Attack-surface risk assessment confidence: 8
- Future hardening guidance confidence: 7
- Scope alignment confidence: 9
- Existing functionality awareness: 8
- Existing tooling/pattern leverage: 8
- Functional completeness confidence: 8
- Pattern correctness confidence: 8
- Documentation coverage confidence: 7


```pr-review-confidence-json
{
  "style_maintainability": 8,
  "repo_convention_adherence": 8,
  "merge_conflict_detection": 7,
  "security_vulnerability_detection": 9,
  "injection_risk_detection": 9,
  "attack_surface_risk_assessment": 8,
  "future_hardening_guidance": 7,
  "scope_alignment": 9,
  "duplication_awareness": 8,
  "tooling_pattern_leverage": 8,
  "functional_completeness": 8,
  "pattern_correctness": 8,
  "documentation_coverage": 7
}
```
